src/syntax/ppx_sedlex.ml

```diff
@@ -253,24 +264,225 @@ let gen_recflag auto =
+        @ List.map (fun ({ actions; _ } : Sedlex.final_case) -> actions) finals
+      in
+      let action2cases = Hashtbl.create n in
+      List.iteri
+        (fun i actions ->
+          List.iter
+            (fun action ->
+              try
+                let offsets = Hashtbl.find action2cases action in
+                Hashtbl.replace action2cases action (i :: offsets)
+              with Not_found -> Hashtbl.add action2cases action [i])
+            actions)
+        cases;
+      let counter = ref 0 in
+      let alias2offset = Hashtbl.create n in
+      let cases2offset = Hashtbl.create 31 in
+      Hashtbl.iter
+        (fun action offsets ->
+          try
+            let i = Hashtbl.find cases2offset offsets in
+            Hashtbl.add alias2offset action i
+          with Not_found ->
+            Hashtbl.add cases2offset offsets !counter;
+            Hashtbl.add alias2offset action !counter;
+            incr counter)
+        action2cases;
+      Some (!counter, alias2offset)
+
+let gen_cset ~loc x char_set =
+  let interval a b =
+    [%expr
+      [%e eint ~loc a] <= [%e evar ~loc x]
+      && [%e evar ~loc x] <= [%e eint ~loc b]]
+  in
+  match char_set with
+    | (a, b) :: l ->
+        List.fold_left
+          (fun acc (a, b) -> [%expr [%e acc] || [%e interval a b]])
+          (interval a b) l
+    | [] -> assert false
+
+let gen_trace lexbuf traces i = function
+  | None -> []
+  | Some (offsets_num, action_offsets) ->
+      let loc = default_loc in
+      let initial, trans, finals = traces.(i) in
+      let offset_array =
+        value_binding ~loc
+          ~pat:[%pat? __sedlex_offsets]
+          ~expr:(pexp_array ~loc (List.init offsets_num (fun _ -> [%expr -1])))
+      in
+      let find_offset_idx action = Hashtbl.find action_offsets action in
+      let aux_fun =
+        let gen_action e offset_idx =
+          [%expr
+            __sedlex_offsets.([%e eint ~loc offset_idx]) <- __sedlex_pos;
+            [%e e]]
+        in
+        let unreachable_case =
+          case ~lhs:[%pat? _] ~guard:None ~rhs:[%expr assert false]
+        in
+        let trans_cases =
+          let dup_case = Hashtbl.create (List.length trans) in
+          List.iter
+            (fun { Sedlex.curr_state; curr_node; prev_state; _ } ->
+              let key = (curr_state, curr_node, prev_state) in
+              if Hashtbl.mem dup_case key then
+                Hashtbl.replace dup_case key false
+              else Hashtbl.add dup_case key true)
+            trans;
+          List.map
+            (fun {
+                   Sedlex.curr_state;
+                   curr_node;
+                   prev_state;
+                   prev_node;
+                   char_set;
+                   actions;
+                 } ->
+              let lhs =
+                ppat_tuple ~loc
+                  [
+                    pint ~loc curr_state;
+                    pint ~loc curr_node;
+                    pint ~loc prev_state;
```

Comment:
is `prev_state` ever needed to choose the action ?
On a trivial example
```ocaml
let rec token buf =
  match%sedlex buf with
    | (Star "a" as a), (Plus "b" as b), "c" ->
        print_alias buf "a" a;
        print_alias buf "b" b;
        token buf
    | eof -> print_endline "EOF"
    | _ -> failwith "Unexpected character"
```
`curr_state` and `curr_node` are enough for choosing the action.
The following
```ocaml
          let __sedlex_code = Sedlexing.lexeme_code buf (__sedlex_pos - 1) in
          (match (__sedlex_curr_state, __sedlex_curr_node,
                   __sedlex_prev_state)
           with
           | (2, 6, 0) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 0 10 __sedlex_rest)
           | (0, 6, 2) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 2 10 __sedlex_rest)
           | (3, 6, 2) -> __sedlex_aux (__sedlex_pos - 1) 2 6 __sedlex_rest
           | (4, 1, 3) -> __sedlex_aux (__sedlex_pos - 1) 3 2 __sedlex_rest
           | (3, 2, 0) ->
               (__sedlex_offsets.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 0 6 __sedlex_rest)
           | (3, 6, 0) -> __sedlex_aux (__sedlex_pos - 1) 0 6 __sedlex_rest
           | (2, 6, 2) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 2 10 __sedlex_rest)
           | (3, 2, 2) ->
               (__sedlex_offsets.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 2 6 __sedlex_rest)
           | (3, 2, 3) ->
               (__sedlex_offsets.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 3 6 __sedlex_rest)
           | (3, 6, 3) -> __sedlex_aux (__sedlex_pos - 1) 3 6 __sedlex_rest
           | (0, 10, 2) -> __sedlex_aux (__sedlex_pos - 1) 2 10 __sedlex_rest
           | (0, 6, 0) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) 0 10 __sedlex_rest)
           | (2, 10, 2) -> __sedlex_aux (__sedlex_pos - 1) 2 10 __sedlex_rest
           | (0, 10, 0) -> __sedlex_aux (__sedlex_pos - 1) 0 10 __sedlex_rest
           | (2, 10, 0) -> __sedlex_aux (__sedlex_pos - 1) 0 10 __sedlex_rest
           | _ -> assert false) in
```

can be rewritten into  

```ocaml
          let __sedlex_code = Sedlexing.lexeme_code buf (__sedlex_pos - 1) in
          (match (__sedlex_curr_state, __sedlex_curr_node) with
           | (2, 6) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 10
                  __sedlex_rest)
           | (0, 6) ->
               (__sedlex_offsets.(2) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 10
                  __sedlex_rest)
           | (3, 6) ->
               __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 6
                 __sedlex_rest
           | (4, 1) ->
               __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 2
                 __sedlex_rest
           | (3, 2) ->
               (__sedlex_offsets.(1) <- __sedlex_pos;
                __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 6
                  __sedlex_rest)
           | (0, 10) ->
               __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 10
                 __sedlex_rest
           | (2, 10) ->
               __sedlex_aux (__sedlex_pos - 1) __sedlex_prev_state 10
                 __sedlex_rest
           | _ -> assert false) in
```
if prev_state is sometime useful, we could include it inside a guard only when needed.

Reply:
The `prev_state` is still useful. It's a good idea to make it a guard.

Reply:
I made a test and found that `prev_state` guards show up frequently, so I decide not to change it.